subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -1427,23 +1286,49 @@ public void requiredBy(Provider provider) {
         }
     }
 
-    private static class ServicesWithTypeLookup implements Transformer<ServiceProvider, Provider> {
-        private final Class<?> elementClass;
+    private static boolean isSatisfiedBy(Type expected, Type actual) {
+        if (expected == actual) {
```

Comment:
:o: Should it be `.equals`?

Reply:
It's mostly for `Class<?>`, which has identity semantics, but yeah `equals()` would work too, so I'll replace it for clarity.